Sources/Classes/NimbleLoyalty.swift

```diff
@@ -5,27 +5,136 @@
 //  Created by David Bui on 16/03/2023.
 //
 
-import Foundation
+import AuthenticationServices
 
-public final class NimbleLoyalty {
+public enum NimbleLoyaltyError: Error {
+
+    case alreadyAuthenticated
+    case clientIdEmpty
+    case clientSecretEmpty
+    case failToCreateSignInURL
+    case failToCreateCallbackURL
+    case failToAuthenticate
+    case failToGetAccessToken(String?)
+    case failToStartASWebAuthenticationSession
+}
+
+extension NimbleLoyaltyError: LocalizedError {
+
+    public var errorDescription: String? {
+        switch self {
+        case .alreadyAuthenticated:
+            return "You've already been authenticated."
+        case .clientIdEmpty:
+            return "Client Id is empty"
+        case .clientSecretEmpty:
+            return "Client Secret is empty"
+        case .failToCreateSignInURL:
+            return "Could not create the sign in URL."
+        case .failToCreateCallbackURL:
+            return "Could not create the callback URL."
+        case .failToAuthenticate:
+            return "An error occurred when attempting to sign in."
+        case let .failToGetAccessToken(error):
+            return "Failed to exchange access code for tokens: \(error ?? "Unknown")"
+        case .failToStartASWebAuthenticationSession:
+            return "Failed to start ASWebAuthenticationSession"
+        }
+    }
+}
+
+public final class NimbleLoyalty: NSObject {
 
     public static let shared = NimbleLoyalty()
 
     private let keyChain = Keychain.default
+    private let authenticationRepository = AuthenticationRepository()
 
-    private init() {}
+    private override init() {}
 
     public func setClientId(_ clientId: String) {
         guard !clientId.isEmpty else {
-            fatalError("Client Id must not be empty")
+            fatalError(NimbleLoyaltyError.clientIdEmpty.localizedDescription)
         }
         try? keyChain.set(clientId, for: .clientId)
     }
 
     public func setClientSecret(_ clientSecret: String) {
         guard !clientSecret.isEmpty else {
-            fatalError("Client Secret must not be empty")
+            fatalError(NimbleLoyaltyError.clientSecretEmpty.localizedDescription)
         }
         try? keyChain.set(clientSecret, for: .clientSecret)
     }
 }
+
+//MARK: Public functions
+
+extension NimbleLoyalty {
+
+    public func isAuthenticated() -> Bool {
+        guard let token: KeychainToken = try? keyChain.get(.userToken) else {
+            return false
+        }
+        return true
+    }
+
+    public func authenticate(_ completion: @escaping (Result<Void, NimbleLoyaltyError>) -> Void) {
```

Comment:
Please create an `AuthenticationSessionManager` to handle these logic instead of putting everything in `NimbleLoyalty` class

Reply:
We should have public APIs in this class